File: fe/src/services/productService.ts
import type { CreateProductInput, Product } from '@/types/pos'
import { apiRequest } from '@/services/apiClient'

export type PageResponse<T> = {
  items: T[]
  page: number
  size: number
  hasNext: boolean
}

export const productService = {
  list(opts?: { deleted?: boolean }): Promise<Product[]> {
    const params = new URLSearchParams()
    if (opts?.deleted) params.set('deleted', 'true')
    const qs = params.toString()
    return apiRequest<Product[]>(`/api/products${qs ? `?${qs}` : ''}`)
  },
  page(opts: {
    q?: string
    deleted?: boolean
    priceMin?: number | null
    priceMax?: number | null
    sortBy?: 'id' | 'price' | 'name' | 'updatedAt'
    sortDir?: 'asc' | 'desc'
    page?: number
    size?: number
  }): Promise<PageResponse<Product>> {
    const params = new URLSearchParams()
    const q = (opts.q ?? '').trim()
    if (q) params.set('q', q)
    if (opts.deleted) params.set('deleted', 'true')
    if (opts.priceMin != null) params.set('priceMin', String(opts.priceMin))
    if (opts.priceMax != null) params.set('priceMax', String(opts.priceMax))
    if (opts.sortBy) params.set('sortBy', opts.sortBy)
    if (opts.sortDir) params.set('sortDir', opts.sortDir)
    params.set('page', String(Math.max(0, Math.floor(opts.page ?? 0))))
    params.set('size', String(Math.max(1, Math.floor(opts.size ?? 20))))
    return apiRequest<PageResponse<Product>>(`/api/products/page?${params.toString()}`)
  },
  /** Tìm theo tên hoặc barcode (server), mặc định tối đa 5 kết quả. */
  search(q: string, limit = 5, opts?: { deleted?: boolean }): Promise<Product[]> {
    const params = new URLSearchParams()
    params.set('q', q.trim())
    params.set('limit', String(limit))
    if (opts?.deleted) params.set('deleted', 'true')
    return apiRequest<Product[]>(`/api/products/search?${params.toString()}`)
  },
  getById(id: number): Promise<Product> {
    return apiRequest<Product>(`/api/products/${id}`)
  },
  getByBarcode(barcode: string): Promise<Product> {
    return apiRequest<Product>(`/api/products/by-barcode/${encodeURIComponent(barcode)}`)
  },
  create(input: CreateProductInput): Promise<Product> {
    return apiRequest<Product>('/api/products', { method: 'POST', body: JSON.stringify(input) })
  },
  update(id: number, input: CreateProductInput): Promise<Product> {
    return apiRequest<Product>(`/api/products/${id}`, {
      method: 'PUT',
      body: JSON.stringify(input),
    })
  },
  delete(id: number): Promise<void> {
    return apiRequest<void>(`/api/products/${id}`, { method: 'DELETE' })
  },
  restore(id: number): Promise<Product> {
    return apiRequest<Product>(`/api/products/${id}/restore`, { method: 'PUT' })
  },
}

File: be/src/main/java/com/example/be/service/ProductService.java
package com.example.be.service;

import com.example.be.dto.request.product.ProductCreateRequest;
import com.example.be.dto.request.product.ProductUpdateRequest;
import com.example.be.dto.response.product.ProductResponse;
import org.springframework.data.domain.Slice;
import org.springframework.data.domain.Sort;

import java.util.List;

public interface ProductService {
    List<ProductResponse> list(boolean deleted);

    /**
     * Tìm theo tên hoặc barcode (LIKE, không phân biệt hoa thường), tối đa {@code limit} bản ghi.
     */
    List<ProductResponse> search(String q, int limit, boolean deleted);

    /**
     * Lọc + phân trang dạng slice (không có tổng số bản ghi) cho infinity scroll.
     */
    Slice<ProductResponse> page(
            String q,
            boolean deleted,
            Integer priceMin,
            Integer priceMax,
            int page,
            int size,
            Sort sort
    );

    ProductResponse getById(Long id);

    ProductResponse getByBarcode(String barcode);

    ProductResponse create(ProductCreateRequest request);

    ProductResponse update(Long id, ProductUpdateRequest request);

    void delete(Long id);

    ProductResponse restore(Long id);
}

